lib/core/src/Cardano/Wallet/Primitive/Model.hs

```diff
@@ -594,40 +631,24 @@ applyOurTxToUTxO !slotNo !blockHeight !tx !prevUTxO = do
                 Just $ distance totalIn totalOut
             (_, Incoming) ->
                 Nothing
-    return if
-        | hasKnownWithdrawal || prevUTxO /= ourNextUTxO ->
-            let adaSpent = TB.getCoin spent
-                adaReceived = TB.getCoin received
-                dir = if adaSpent > adaReceived then Outgoing else Incoming
-                amount = distance adaSpent adaReceived
-            in
-            Just ((tx {fee = actualFee dir}, mkTxMeta amount dir), ourNextUTxO)
-        | otherwise ->
-            Nothing
-  where
-    mkTxMeta :: Coin -> Direction -> TxMeta
-    mkTxMeta amount dir = TxMeta
-        { status = InLedger
-        , direction = dir
-        , slotNo
-        , blockHeight
-        , amount = amount
-        , expiry = Nothing
-        }
-
--- | Get the change UTxO
---
--- NOTE
--- We _safely_ discard the state here because we aren't intending to
--- discover any new addresses through this operation. In practice, we
--- can only discover new addresses when applying blocks. The state is
--- therefore use in a read-only mode here.
-changeUTxO
-    :: IsOurs s Address
-    => Set Tx
-    -> s
-    -> UTxO
-changeUTxO pending = evalState $
-    mconcat <$> mapM
-        (filterByAddressM isOurAddress . utxoFromUnvalidatedTx)
-        (Set.toList pending)
+    in if hasKnownWithdrawal || prevUTxO /= ourNextUTxO
+        then Just ((tx {fee = actualFee dir}, txmeta), ourNextUTxO)
+        else Nothing
+
+ourWithdrawalSumFromTx
+    :: IsOurs s RewardAccount
+    => s -> Tx -> Coin
+ourWithdrawalSumFromTx s tx
+    -- If a transaction has failed script validation, then the ledger rules
+    -- require that applying the transaction shall have no effect other than
+    -- to fully spend the collateral inputs included within that transaction.
+    --
+    -- Therefore, any reward withdrawals included in such a transaction should
+    -- also have no effect.
+    --
+    | failedScriptValidation tx =
+        Coin 0
+    | otherwise =
+        F.foldMap snd
+        . filter (ours s . fst)
```

Comment:
I wonder if we could be more performant here :thinking:

Reply:
You mean like `foldMap` on the map directly, without `toList` and `filter` in order to avoid building intermediate maps?

Reply:
possibly, Map has Foldable instance ... we have Map.filterWithKey. just food for thought